LocalDatabase vends separate validating BatchStore

cmasone-attic opened the pull request:
This patch modifies LocalDatabase so that it no longer swaps out
its embedded ValueStore during Pull(). The reason it was doing this
is that Pull() injects chunks directly into a Database, without
doing any work on its own to ensure correctness. For LocalDatabase,
WriteValue() performs de-facto validation as it goes, so it does not
need this additional validation in the general case. To address the
former wtithout impacting the latter, we were making LocalDatabase
swap out its ValueStore() during Pull(), replacing it with one that
performs validation.

This led to inconsistencies, seen in issue #2598. Collections read
from the DB _before_ this swap could have weird behavior if they
were modified and written after this swap.

The new code just lazily creates a BatchStore for use during Pull().

Fixes #2598

cmasone-attic committed:
LocalDatabase vends separate validating BatchStore

This patch modifies LocalDatabase so that it no longer swaps out
its embedded ValueStore during Pull(). The reason it was doing this
is that Pull() injects chunks directly into a Database, without
doing any work on its own to ensure correctness. For LocalDatabase,
WriteValue() performs de-facto validation as it goes, so it does not
need this additional validation in the general case. To address the
former wtithout impacting the latter, we were making LocalDatabase
swap out its ValueStore() during Pull(), replacing it with one that
performs validation.

This led to inconsistencies, seen in issue #2598. Collections read
from the DB _before_ this swap could have weird behavior if they
were modified and written after this swap.

The new code just lazily creates a BatchStore for use during Pull().

Fixes #2598

cmasone-attic commented:
PTAL @rafael-atticlabs 
FYI @ehalpern

cmasone-attic commented:
Jenkins: perf this

rafael-atticlabs commented on go/datas/local_database.go:
```diff
@@ -29,36 +30,40 @@ func (ldb *LocalDatabase) GetDataset(datasetID string) Dataset {
-		ldb.ValueStore = types.NewValueStore(bs)
-		ldb.rt = bs
+func (ldb *LocalDatabase) validatingBatchStore() types.BatchStore {
+	if ldb.vbs == nil {
```
ha. nice.

rafael-atticlabs commented on cmd/noms/noms_sync_test.go:
```diff
@@ -78,6 +81,49 @@ func (s *nomsSyncTestSuite) TestSync() {
+	sourceDB := datas.NewDatabase(chunks.NewLevelDBStore(s.LdbDir, "", 1, false))
+	// Create dataset "src1", which has a lineage of two commits.
+	source1 := sourceDB.GetDataset("src1")
+	source1, err := sourceDB.CommitValue(source1, types.Number(42))
```
the new api is nice.

cmasone-attic committed:
Tweak localBatchStore API for more correctness

Added a new little method to call that allows a localBatchStore
to be flushed and destroyed without closing the backing ChunkStore,
since that's what LocalDatabase needs -- and its the only caller
of these methods.
